File: src/Infrastructure/Vpiska.Firebase/Entry.cs
using FirebaseAdmin;
using Google.Apis.Auth.OAuth2;
using Google.Cloud.Storage.V1;
using Microsoft.Extensions.DependencyInjection;
using Vpiska.Domain;

namespace Vpiska.Firebase
{
    public static class Entry
    {
        public static void AddFirebase(this IServiceCollection services)
        {
#if DEBUG
            const string path = "../Infrastructure/Vpiska.Firebase/firebase.json";
#else
            const string path = "firebase.json";
#endif

            services.AddSingleton(FirebaseApp.Create(new AppOptions()
            {
                Credential = GoogleCredential.FromFile(path)
            }));

            services.AddSingleton(StorageClient.Create(GoogleCredential.FromFile(path)));
            services.AddTransient<INotificationService, NotificationService>();
            services.AddTransient<IFileStorage, FileStorage>();
        }
    }
}

File: src/Vpiska.Infrastructure/Vpiska.Infrastructure.Firebase/Entry.fs
namespace Vpiska.Infrastructure.Firebase

open System.Runtime.CompilerServices
open FirebaseAdmin
open Google.Apis.Auth.OAuth2
open Google.Cloud.Storage.V1
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection

type FirebaseSettings = { BucketName: string }

[<Extension>]
type Entry() =
    
    [<Extension>]
    static member AddFirebase(services: IServiceCollection, firebaseSection: IConfigurationSection) =
        #if DEBUG
        let path = "../Vpiska.Infrastructure/Vpiska.Infrastructure.Firebase/settings.json"
        #else
        let path = "settings.json"
        #endif
        let settings = { BucketName = firebaseSection.["BucketName"] }
        let firebaseApp = FirebaseApp.Create(AppOptions(Credential = GoogleCredential.FromFile(path)))
        let storageClient = StorageClient.Create(GoogleCredential.FromFile(path))
        services.AddSingleton(settings) |> ignore
        services.AddSingleton(firebaseApp) |> ignore
        services.AddSingleton(storageClient) |> ignore
    